design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,777 @@
+# installed providers.
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - exampleproviderconfigs
+  verbs: ["*"]
+```
+
+The `crossplane-edit` role is identical to `crossplane-admin`, sans the ability
+to grant others access by managing role bindings.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-edit
+rules:
+# Crossplane editors have access to view events.
+- apiGroups: [""]
+  resources: [events]
+  verbs: [get, list, watch]
+# Crossplane editors must create provider credential secrets, and may need to
+# read or otherwise interact with connection secrets.
+- apiGroups: [""]
+  resources: [secrets]
+  verbs: ["*"]
+# Crossplane editors may see which namespaces exist, but not edit them.
+- apiGroups: [""]
+  resources: [namespaces]
+  verbs: [get, list, watch]
+# Crossplane editors have full access to built in Crossplane types.
+- apiGroups:
+  - apiextensions.crossplane.io
+  - pkg.crossplane.io
+  resources: ["*"]
+  verbs: ["*"]
+# Crossplane editors have full access to all of the composite resources and
+# claims it defines.
+- apiGroups: [xr.example.org]
+  resources:
+  - examplecomposites
+  - examplecomposites/status
+  - exampleclaims
+  - exampleclaims/status
+  verbs: ["*"]
+# Crossplane viewers have full access to all resources defined by installed
+# providers.
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - exampleproviderconfigs
+  verbs: ["*"]
+```
+
+The `crossplane-view` role allows read-only access to all Crossplane resources.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane-view
+rules:
+# Crossplane viewers have access to view events.
+- apiGroups: [""]
+  resources: [events]
+  verbs: [get, list, watch]
+# Crossplane viewers may see which namespaces exist.
+- apiGroups: [""]
+  resources: [namespaces]
+  verbs: [get, list, watch]
+# Crossplane viewers have read-only access to built in Crossplane types.
+- apiGroups:
+  - apiextensions.crossplane.io
+  - pkg.crossplane.io
+  resources: ["*"]
+  verbs: [get, list, watch]
+# Crossplane viewers have read-only access to all of the composite resources and
+# claims it defines.
+- apiGroups: [xr.example.org]
+  resources:
+  - examplecomposites
+  - examplecomposites/status
+  - exampleclaims
+  - exampleclaims/status
+  verbs: [get, list, watch]
+# Crossplane viewers have read-only access to all resources defined by installed
+# providers.
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - exampleproviderconfigs
+  verbs: [get, list, watch]
+```
+
+The RBAC manager creates `edit` and `view` cluster roles for each namespace.
+These cluster roles are 'namespace aligned' - they are intended to grant access
+to a particular namespace via a `RoleBinding` - but not namespace scoped. This
+is because (while these examples are flattened into single roles) they use role
```

Comment:
Is the namespace-specific aggregation the only reason? Do you expect those roles to have cluster-scoped permissions at some point? (I'm not sure if we discussed this, probably did). If you don't expect cluster-scoped permissions to be added in the future, maybe RBAC manager could do the aggregation manually, i.e. look at annotations on the `Namespace` and update `Role`s in that namespace.

Reply:
> Is the namespace-specific aggregation the only reason? Do you expect those roles to have cluster-scoped permissions at some point.

It is, and I don't.

> maybe RBAC manager could do the aggregation manually

I was on the fence about this too. On one hand I appreciate and share the desire to use a namespaced `Role` to represent a role that applies to a specific namespace. On the other hand I'm reluctant to 'reinvent' aggregation functionality that already exists. I might sketch out what it would look like if we used a `Role` after I address the other comments in this doc.

Reply:
Take a look at the latest commit - I decided to go with having the RBAC manager manage namespaced roles by emulating/approximating an opinionated form of cluster role aggregation. I went back and forth a lot, but it seems like it's not really any more difficult to manually aggregate RBAC role rules (as opposed to manually managing aggregation rules), and it feels a little more 'idiomatic' than creating a cluster role for each namespace.

Reply:
Nice! I love that we're able to do that. There will be some complexity added to the controller code but I believe that's manageable given the cost of the alternative. Thanks!